File: 入社テスト2023/Q2/Main.java
import java.util.*;

public class Main {
	
	private class Rule {
		public int target; // N - Target
		public int maxSpeed; // X - Coding speed
		public int reduceAfterDev; // F - Reduce speed by F after writing code
		public int gainAfterSleep; // S - Gain speed by S after sleep for 3 hours
	
		public Rule(int target, int maxSpeed,int reduceAfterDev, int gainAfterSleep) {
			this.target = target;
			this.maxSpeed = maxSpeed;
			this.reduceAfterDev = reduceAfterDev;
			this.gainAfterSleep = gainAfterSleep;
		}
	}

	private enum State {
		ROOT,
		SLEEP,
		DEVELOP
	}

	private Rule getInput() {
		Scanner scanner = new Scanner(System.in);
		String line = scanner.nextLine();
		scanner.close();

		String parts[] = line.split(" ");
		return new Rule(
			Integer.parseInt(parts[0]),
			Integer.parseInt(parts[1]),
			Integer.parseInt(parts[2]),
			Integer.parseInt(parts[3])
		);
	}

	private int getHours(Rule rule, int hoursSpent, int currentSpeed, int linesLeftToWrite, State state) {
		if (currentSpeed <= 0) { // Force sleep
			hoursSpent += 3;
			currentSpeed = rule.gainAfterSleep;
		} else if (currentSpeed >= rule.maxSpeed) { // Force develop
			hoursSpent += 1;
			linesLeftToWrite -= currentSpeed;
			currentSpeed = Math.max(currentSpeed - rule.reduceAfterDev, 0);
		} else if (state == State.SLEEP) { // sleep
			hoursSpent += 3;
			currentSpeed = Math.min(currentSpeed + rule.gainAfterSleep, rule.maxSpeed);
		} else if (state == State.DEVELOP) { // develop
			hoursSpent += 1;
			linesLeftToWrite -= currentSpeed;
			currentSpeed = Math.max(currentSpeed - rule.reduceAfterDev, 0);
		}

		if (linesLeftToWrite <= 0) {
			return hoursSpent;
		}

		return Math.min(
			getHours(rule, hoursSpent, currentSpeed, linesLeftToWrite, State.SLEEP),
			getHours(rule, hoursSpent, currentSpeed, linesLeftToWrite, State.DEVELOP)
		);
	}

	public Main() {
		Rule rule = getInput();

		System.out.println(getHours(rule, 0, rule.maxSpeed, rule.target, State.ROOT));
	}

	public static void main(String[] args) {
		new Main();
	}

}

File: 入社テスト2023/Q2/main.cpp
#include <cmath>
#include <fstream>
#include <iostream>
#include <vector>


struct Rule {
	int target; // N - Target
	int max_speed; // X - Coding speed
	int reduce_after_dev; // F - Reduce speed by F after writing code
	int gain_after_sleep; // S - Gain speed by S after sleep for 3 hours
};

enum class State {
	kRoot,
	kSleep,
	kDevelop
};

Rule get_input() {
	Rule rule { 0, 0, 0, 0 };
	std::cin >> rule.target >> rule.max_speed >> rule.reduce_after_dev >> rule.gain_after_sleep;
	return rule;
}

Rule get_input_from_file(const char *file_path) {
	std::ifstream fs(file_path);
	
	if (fs.is_open()) {
		std::cin.rdbuf(fs.rdbuf());
		return get_input();
	}
	return Rule { 0, 0, 0, 0 };
}

int get_hours(const Rule &rule, int hours_spent, int current_speed, int lines_left_to_write, State state) {
	if (current_speed <= 0) { // Force Sleep
		hours_spent += 3;
		current_speed = rule.gain_after_sleep;
	} else if (current_speed >= rule.max_speed) { // Force Develop
		hours_spent += 1;
		lines_left_to_write -= current_speed;
		current_speed = std::max(current_speed - rule.reduce_after_dev, 0);
	} else if (state == State::kSleep) { // Sleep
		hours_spent += 3;
		current_speed = std::min(current_speed + rule.gain_after_sleep, rule.max_speed);
	}else if (state == State::kDevelop) { // Develop
		hours_spent += 1;
		lines_left_to_write -= current_speed;
		current_speed = std::max(current_speed - rule.reduce_after_dev, 0);
	}

	if (lines_left_to_write <= 0) {
		return hours_spent;
	}

	return std::min(
		get_hours(rule, hours_spent, current_speed, lines_left_to_write, State::kSleep),
		get_hours(rule, hours_spent, current_speed, lines_left_to_write, State::kDevelop)
	);
}

int main() {
	Rule rule = get_input();
	// Rule rule =	get_input_from_file("tc.txt");

	std::cout << get_hours(rule, 0, rule.max_speed, rule.target, State::kRoot) << std::endl;
}